jetbrains-core/src/software/aws/toolkits/jetbrains/settings/AwsSettings.kt

```diff
@@ -30,6 +32,17 @@ class AwsSettings : PersistentStateComponent<AwsConfiguration> {
             state.promptedForTelemetry = value
         }
 
+    val clientId: UUID
+        get() {
+            synchronized(_lock) {
+                if (state.clientId == null) {
+                    state.clientId = UUID.randomUUID()
+                }
+            }
+
+            return state.clientId!!
```

Comment:
* I switched to the `@Synchronized` attribute, but the compiler still isn't able to determine that `state.clientId` will always be non-null.
* What is the value in falling back on a transient random UUID? A transient ID does not follow the semantics of `clientId`. Better to log no data than incorrect data.
* Isn't this scenario (we can reason that `state.clientId` can't be null, but the compiler can't) exactly what the `!!` operator is designed for?

Reply:
You can write it as the following to remove it:

```
return if (state.clientId == null) {
    UUID.randomUUID().also {
        state.clientId = it
    }
} else {
    state.clientId
}
        
            return state.clientId!!

Reply:
If anything we should not save this in a PersistentStateComponent but instead in the users's registry. Otherwise we run the risk of double counting users between IntelliJ 2018.2 and 2018.3.

Reply:
✔️